lib/std/os/uefi/tables/boot_services.zig

```diff
@@ -32,173 +42,1223 @@ pub const BootServices = extern struct {
+        parent_image: Handle,
+        source: LoadImageSource,
+    ) LoadImageError!Handle {
+        var handle: Handle = undefined;
+
+        switch (self._loadImage(
+            boot_policy,
+            parent_image,
+            if (source == .device_path) source.device_path else null,
+            if (source == .buffer) source.buffer.ptr else null,
+            if (source == .buffer) source.buffer.len else 0,
+            &handle,
+        )) {
+            .success => return handle,
+            .not_found => return Error.NotFound,
+            .invalid_parameter => return Error.InvalidParameter,
+            .unsupported => return Error.Unsupported,
+            .out_of_resources => return Error.OutOfResources,
+            .load_error => return Error.LoadError,
+            .device_error => return Error.DeviceError,
+            .access_denied => return Error.AccessDenied,
+            .security_violation => return Error.SecurityViolation,
+            else => |status| return uefi.unexpectedStatus(status),
+        }
+    }
+
+    pub fn startImage(self: *BootServices, image: Handle) StartImageError!ImageExitData {
+        var exit_data_size: usize = undefined;
+        var exit_data: [*]u16 = undefined;
+
+        const exit_code = switch (self._startImage(
+            image,
+            &exit_data_size,
+            &exit_data,
+        )) {
+            .invalid_parameter => return Error.InvalidParameter,
+            .security_violation => return Error.SecurityViolation,
+            else => |exit_code| exit_code,
+        };
+
+        if (exit_data_size == 0) return .{
+            .code = exit_code,
+            .description = null,
+            .data = null,
+        };
+
+        const description_ptr: [*:0]const u16 = @ptrCast(exit_data);
+        const description = std.mem.sliceTo(description_ptr, 0);
+
+        return ImageExitData{
+            .code = exit_code,
+            .description = description,
+            .data = exit_data[description.len + 1 .. exit_data_size],
+        };
+    }
+
+    pub const ExitError = uefi.UnexpectedError || error{
+        InvalidParameter,
+    };
+
+    /// `data` may be a [*:0]const u16 immediately following additional data of
+    /// any format. The string is a description that the caller may use to further
+    /// indicate the reason for the image’s exit. `data` is only valid if `status`
+    /// is something other than `.success`.
+    pub fn exit(
+        self: *BootServices,
+        handle: Handle,
+        status: Status,
+        data: ?[]const u8,
+    ) ExitError!void {
+        switch (self._exit(
+            handle,
+            status,
+            if (data) |d| d.len else 0,
+            if (data) |d| @alignCast(@ptrCast(d.ptr)) else null,
+        )) {
+            .success => {},
+            .invalid_parameter => return error.InvalidParameter,
+            else => |exit_status| return uefi.unexpectedStatus(exit_status),
+        }
+    }
```

Comment:
Could this guarantee alignment safety at compile-time?

```suggestion
    /// `data` may be a [*:0]const u16 immediately following additional data of
    /// any format. The string is a description that the caller may use to further
    /// indicate the reason for the image’s exit. `data` is only valid if `status`
    /// is something other than `.success`.
    pub fn exit(
        self: *BootServices,
        handle: Handle,
        status: Status,
        data: ?[]const align(2) u8,
    ) ExitError!void {
        switch (self._exit(
            handle,
            status,
            if (data) |d| d.len else 0,
            if (data) |d| @ptrCast(d.ptr) else null,
        )) {
            .success => {},
            .invalid_parameter => return error.InvalidParameter,
            else => |exit_status| return uefi.unexpectedStatus(exit_status),
        }
    }
```

Reply:
Additionally, I'm not really sure that I like that this interface uses a slice of bytes rather than `u16`, when the principal use of this is to return the string, since it makes this sort of invocation of this API use a `@ptrcast`. However, since you can pass binary data after the null-terminated string, I guess that makes sense? The only alternative I can see would be to split into `exit` (taking `[:0]const u16`) and `exitBytes` (`[]const align(2) u8`), which is not really ideal either.

Reply:
No, I think accepting two pointers isn't that great a solution, could easily introduce a footgun, and isn't obvious from the function signature that they must be contiguous. I think either we should have two different wrapper functions, or status quo is sufficient.

Reply:
Given the binary blob after the string in exit data is vendor-specific, it may be worth splitting this into two separate functions: `exit` which just takes a `[:0]const u16` message string, and `exitWithData` which accepts an `[]align(2) const u8` and we just state the assumption that the data starts with a u16 string.

Reply:
i like that solution :) i'll do that